### DISCUSSION TITLE
Forecast without retrain the model

### DISCUSSION THREAD

**Discussion by raphaccf** [pycaret/pycaret#1969] [Q&A] [answered]
Hi,
I am enjoying working with TS Pycaret.
However, I have a question about the `finalize_model()` function, maybe it's silly, sorry in advance.
Does this function really retrain, in the sense of changing the parameters of the models that went through the entire modeling pipeline and got tunned? It sounds for me that information/data for evaluating is leaking, if it changes the parameters.
Or does it just put holdout as an input (append to the "y"/target variable), to forecast  n_steps ahead, after the last date of the data set?

In this sense, for example, if I have new data and would just like to make a forecast (append more data to y), without retrain the model (in the senso of changing parameters) how could I proceed?

Thanks

**Comment by ngupta23** [comment]
No, `finalize_model` will not change the model hyperparameters. It picks the hyperparameters from the model that you pass to it and then retrains using all the data (train + test), but the hyperparameters remain the same. This allows you to make "true" future forecasts which are not available at that point in time.

**Comment by ngupta23** [comment, marked as answer]
> In this sense, for example, if I have new data and would just like to make a forecast (append more data to y), without retrain the model (in the senso of changing parameters) how could I proceed?

You are talking about online learning. Here is a guide to this.

https://github.com/pycaret/pycaret/discussions/1767

**Reply by raphaccf** [reply]
Hi, Nikhil, thanks for your response.
I have created, using auto.arima, an example that shows that parameters and the model change (not the type/method of the model, such as auto.arima, but they are two differente auto.arima's), after "finalize_model()". I used sktime and tried some fuctions, that by their documentation, would prevent for updating parameters (similar to what is "online learning" that you mentioned), and changing the model that was estimated, competed, and tunned (in a standard model pipeline in TS Pycaret). However, it seems that this funciton from sktime is not working well (avoid updating parameters). Here is the link for the notebook:

https://github.com/raphaccf/other/blob/446851fa578cd01517b0f8b1886b72cd50597cc9/TSPycaret_study_finalize_modelx.ipynb

Thanks!

**Reply by ngupta23** [reply]
AutoARIMA is an exception to this rule because of the way in which it works. All models maintain the hyperparameters except AutoARIMA. If you want to maintain the parameters, you should use ARIMA instead of AutoARIMA.

**Reply by raphaccf** [reply]
I have made a test with "ets", and it also changed parameters (coefs) from the model creation to the finalize_model() function (tunning or not this method of model, such change is observed, since number of obsevatios shows that it changed from 138 to 144, for airline dataset). Moreover, It really seems that sktime function `.update(y=get_config('y_test'), update_params=False)  ` is not working, when set `update_params=False` - an online learning with retrain the model (change parameters). When it is set to True, then you have the same results as finalize_model(), e.g., the model is retrained and you have got new parameters/different model (but same method, ets). Besides being from Pycaret Regression, a made a test with `lightgbm_cds_dt `. When using `.get_params()` (tunning or not), hyperparameters remains the same (I do not know how to retriev paramaters from lgbm). However, using `.update(y=get_config('y_test'), update_params=False)  ` the results are similar to the others - if update_params=True we have got the same results as finalize_model(), but whe it is False, the function is no appending "y_test" and making predictions after it without changing parameters.

**Comment by ngupta23** [comment]
Can you provide me a reproducible example with data for `ets`? I can have a look after that. Thanks!

**Reply by raphaccf** [reply]
Sure!!

```
`from pycaret.time_series import *
from pycaret.datasets import get_data
data = get_data('airline', verbose=True)

setup(data, 
      fh = 6, 
      fold = 3, 
      fold_strategy='expanding',
      enforce_pi=False,
      seasonal_period=12,
      session_id = 123)

ets = create_model('ets',cross_validation=False)`
```

`predict_model(ets) `

```
final_ets = finalize_model(ets)
predict_model(final_ets)
```

`ets.summary()`

```
### different parameters (including number of observations) from the first model (ets)
final_ets.summary()
```


```
### tunning
tuned_ets =tune_model(ets, n_iter=5, search_algorithm='random',
                         choose_better=True,
                         tuner_verbose=False,
                         optimize='MAE',
                         fold=1)
```


```
### the same parameters as the ets model
tuned_ets.summary()
```

```
### the same parameters as the ets_final model
final_tuned_ets = finalize_model(tuned_ets)
final_tuned_ets.summary()
```

`predict_model(ets) `

```
updated_model = ets.update(y=get_config('y_test'), update_params=False)
updated_model.predict()
```

```
### if we update we get the same results as finalize_model(), so parameters are changing 
updated_model2 = ets.update(y=get_config('y_test'), update_params=True)
updated_model2.predict()
```
[pdf_ets_tspycaret.pdf](https://github.com/pycaret/pycaret/files/7816547/pdf_ets_tspycaret.pdf)

**Reply by ngupta23** [reply]
@raphaccf I see the issue with the way you are interpreting this and your flow. There are many points to discuss so I will try to list one by one.

```python
### different parameters (including number of observations) from the first model (ets)
final_ets.summary()
```

* Between `create_model`/`tune_model` and `finalize_model`, what changes is the data used for training.  `create_model`/`tune_model` use the training dataset only, while `finalize_model` uses the entire dataset (train + test). What remains same between `create_model`/`tune_model` and `finalize_model` are the model **hyperparameters** (NOTE: model **hyperparameters** are different from model **parameters**).
* When I say model **hyperparameters** remain same, I mean that if in `create_model`, if you have `seasonal_period` of 12 (corresponding to **parameters**  `initial_seasonal.0` till `initial_seasonal.11`), in the final model also you will have you have seasonal period of 12.
* However the value of these **parameters**  `initial_seasonal.0` till `initial_seasonal.11` will get updated when you finalize the model since there is more data (train + test) when you are finalizing. The values are determined by the best fit to this complete dataset. This is the expected behavior.
  

```python
### tunning
tuned_ets =tune_model(ets, n_iter=5, search_algorithm='random',
                         choose_better=True,
                         tuner_verbose=False,
                         optimize='MAE',
                         fold=1)

### the same parameters as the ets model
tuned_ets.summary()
```

- The reason you get the same parameters as `ets` is because you have used `choose_better = True`. What this does is that if the grid search does not produce a better result than the original model (`ets` in this case, it will return back the original model). Hence by definition, it will be same as `ets`.

```python
updated_model = ets.update(y=get_config('y_test'), update_params=False)
updated_model.predict()

### if we update we get the same results as finalize_model(), so parameters are changing 
updated_model2 = ets.update(y=get_config('y_test'), update_params=True)
updated_model2.predict()
```

- I do not see an issue here either.
- If you use `update_params=False`, then the value of **parameters**  `initial_seasonal.0` till `initial_seasonal.11` will remain the same as that in `ets`
- If you use `update_params=True`, then the value of **parameters**  `initial_seasonal.0` till `initial_seasonal.11` will change. Since what you are doing manually here is the same as what `finalize_model` does, you will get the same results as `finalize_model`.

I hope it makes sense now. Let me know.

**Reply by raphaccf** [reply]
Yes, this make sense. Many thanks!!!
